certora/specs/SocialRecoveryModule.spec

```diff
@@ -1,26 +1,131 @@
-// Module is working as intended.
-rule recoveryModuleCanBeDisabled {
-    env e;
-    address prevModule;
+// This integrity rule verifies that the addGuardianWithThreshold(...) works as expected.
+rule addGuardianWorksAsExpected(env e, address guardian, uint256 threshold, address otherGuardian) {
+
+    // This can be removed once linked list invariant is implemented.
+    require guardianStorageContract.entries[safeContract].count > 0 => guardianStorageContract.entries[safeContract].guardians[SENTINEL()] != 0; // Sentinel should not be zero if there are guardians.
+
+    require guardian != otherGuardian;
+
+    bool otherWasGuardian = currentContract.isGuardian(safeContract, otherGuardian);
+
+    currentContract.addGuardianWithThreshold(e, safeContract, guardian, threshold);
+
+    assert safeContract.isModuleEnabled(currentContract);
+    assert e.msg.sender == safeContract;
+    assert currentContract.isGuardian(safeContract, guardian);
+    assert otherWasGuardian == currentContract.isGuardian(safeContract, otherGuardian);
+}
+
+// This integrity rule verifies that the guardian can always be added considering ideal conditions.
+rule guardianCanAlwaysBeAdded(env e, address guardian, uint256 threshold, address otherGuardian) {
 
     requireSocialRecoveryModuleEnabled();
 
-    safeContract.disableModule@withrevert(e, prevModule, currentContract);
+    require e.msg.value == 0;
+    require threshold > 0;
+    require guardianStorageContract.entries[safeContract].count < max_uint256; // To prevent overflow (Realistically can't reach).
+
+    // This can be removed once linked list invariant is implemented.
+    require guardianStorageContract.entries[safeContract].count > 0 => guardianStorageContract.entries[safeContract].guardians[SENTINEL()] != 0; // Sentinel should not be zero if there are guardians.
+
+    require guardian != 0;
+    require guardian != SENTINEL();
+    require guardian != safeContract;
+    require guardian != otherGuardian;
+
+    require !safeContract.isOwner(guardian);
+    require !currentContract.isGuardian(safeContract, guardian);
+
+    require threshold < guardianStorageContract.countGuardians(safeContract);
+
+    bool otherWasGuardian = currentContract.isGuardian(safeContract, otherGuardian);
+
+    require e.msg.sender == safeContract;
+    currentContract.addGuardianWithThreshold@withrevert(e, safeContract, guardian, threshold);
+    bool isReverted = lastReverted;
+
+    assert !isReverted => currentContract.isGuardian(safeContract, guardian);
+    assert otherWasGuardian == currentContract.isGuardian(safeContract, otherGuardian);
+}
+
+// This integrity rule verifies that the revokeGuardianWithThreshold(...) works as expected.
+rule removeGuardiansWorksAsExpected(env e, address guardian, address prevGuardian, uint256 threshold) {
+
+    address nextGuardian = guardianStorageContract.entries[safeContract].guardians[guardian];
```

Comment:
I liked the `otherGuardian` check (that basically verifies that `revokeGuardianWithThreshold` does not change the status of any address that isn't `guardian`). Can we add that to this rule as well?

Reply:
Done 👍🏾